File: server/app/api/files.py
"""文件服务:解析产物的图片出口(M1.5 的最后一环)+ 上传原件的 PDF 出口。

MinerU 把图/表/公式切成独立的 jpg,markdown 里是 `images/<sha256>.jpg` 相对路径 ——
**没有这个端点,校对页和答案里的图片就是一堆 404 占位框**。

PDF 出口是给校对页做原文对照用的:左边浏览器原生 PDF 阅读器(`#page=N` 可直接跳页),
右边编辑解析出来的 markdown —— 校对的意思就是"对着原件看解析对不对",少了原件就没得对。

只读、只服务 FILE_STORAGE_DIR 之内的路径、图片只允许单层文件名(见 `_safe_name`):
路径穿越(`..%2f`)在这类"按名字取文件"的端点上是最常见的洞。
"""

import re
import uuid

from fastapi import APIRouter
from fastapi.responses import FileResponse

from app.api.deps import SessionDep
from app.config import settings
from app.core.errors import NotFoundError
from app.models import Document
from app.services.exact_qa import storage

router = APIRouter(prefix="/api/files", tags=["files"])

#: MinerU 的图片名是 sha256 + 扩展名;放宽一点允许连字符/下划线,但**不允许任何分隔符**
_SAFE_NAME = re.compile(r"^[A-Za-z0-9._-]{1,128}$")
_CONTENT_TYPES = {".jpg": "image/jpeg", ".jpeg": "image/jpeg", ".png": "image/png"}


def _safe_name(name: str) -> str:
    if not _SAFE_NAME.match(name) or name.startswith("."):
        raise NotFoundError("Image not found", code="image_not_found")
    return name


@router.get("/parses/{document_id}/images/{name}")
async def get_parse_image(document_id: uuid.UUID, name: str) -> FileResponse:
    """`GET /api/files/parses/{document_id}/images/{name}` —— 契约见 schemas/exact_qa.py。"""
    path = storage.images_dir(str(document_id)) / _safe_name(name)
    if not path.is_file():
        raise NotFoundError("Image not found", code="image_not_found")
    media = _CONTENT_TYPES.get(path.suffix.lower(), "application/octet-stream")
    # 内容按 sha256 命名 = 内容寻址,可以放心长缓存
    return FileResponse(path, media_type=media, headers={"Cache-Control": "public, max-age=86400"})


@router.get("/documents/{document_id}/pdf")
async def get_source_pdf(document_id: uuid.UUID, session: SessionDep) -> FileResponse:
    """上传原件(校对页左侧的 PDF 预览)。

    路径不从 URL 拼,而是从 `documents.raw_uri` 取(入库时由 storage 层写的相对路径),
    再断言它落在 FILE_STORAGE_DIR 之内 —— 库里的值也不当可信输入。
    """
    doc = await session.get(Document, document_id)
    if doc is None or not doc.raw_uri:
        raise NotFoundError("Source file not found", code="source_not_found")
    root = settings.storage_path.resolve()
    path = (root / doc.raw_uri).resolve()
    if not path.is_file() or root not in path.parents:
        raise NotFoundError("Source file not found", code="source_not_found")
    # inline 而不是 attachment:要在 iframe 里显示,不是让人下载。
    # 刻意不回传文件名 —— 原名是用户上传的字符串,拼进响应头是没必要的风险
    return FileResponse(
        path,
        media_type="application/pdf",
        headers={"Content-Disposition": "inline", "Cache-Control": "private, max-age=3600"},
    )

File: web/src/domains/document/files.ts
/** 静态产物的 URL 拼装 —— 图片与原件 PDF 走 `<img>` / `<iframe>`,不经过 `apiFetch`,
 *  所以前缀要自己带上(开发环境走 Vite 代理时 `API_BASE` 是空串)。
 *
 *  路由出处:`server/app/api/files.py`。 */

import { API_BASE } from '@/api/client'

/** 切片里那张图的 URL。
 *
 * @param documentId 文档 id(在 `origin_ref.document_id` 里)。
 * @param img payload 里的相对路径,形如 `images/<sha256>.jpg`。
 * @returns 可直接放进 `<img src>` 的地址。
 */
export function figureUrl(documentId: string, img: string): string {
  return `${API_BASE}/api/files/parses/${documentId}/${img}`
}

/** 原件 PDF 的 URL,可选跳到某一页。
 *
 * 用浏览器原生阅读器(与 S1 校对页同一个取舍):S2 需要的只是"翻到第 N 页对着看",
 * 不值得为它引 pdf.js。代价是 bbox 高亮做不了。
 *
 * @param documentId 文档 id。
 * @param page 人看的页号(从 1 起);不给就从第一页开始。
 * @returns 带 `#page=N&view=FitH` 片段的地址。
 */
export function documentPdfUrl(documentId: string, page?: number): string {
  const base = `${API_BASE}/api/files/documents/${documentId}/pdf`
  return page ? `${base}#page=${page}&view=FitH` : base
}
